item_sort_key always uses itemgetter in 3.11

Got a strange bug when upgrading 3.10.0 to 3.11.0 via pip.

A test started failing that used a custom sort function via `item_sort_key`. It seems that setting this kwarg has the effect of enabling sorting using the default `itemgetter` in 3.11.0.

This occurs when using the C encoder implementation.

Python is 2.7.11 on Ubuntu 16.04.

Using this code:

    import simplejson

    def sortit(x):
      return x[1]

    a = {
      "c": "3",
      "b": "1",
      "a": "2"
    }

    print simplejson.dumps(a, item_sort_key=sortit)

On 3.10.0:

    $ python test.py
    {"b": "1", "c": "2", "a": "3"}

And on 3.11.0:

    $ python test.py
    {"a": "2", "b": "1", "c": "3"}

Same here. The related tests are also failing when the C extension is present:

```
======================================================================== FAILURES =========================================================================
________________________________________________________________ TestItemSortKey.test_case ________________________________________________________________

self = <simplejson.tests.test_item_sort_key.TestItemSortKey testMethod=test_case>

    def test_case(self):
        a = {'a': 1, 'c': 5, 'Jack': 'jill', 'pick': 'axe', 'Array': [1, 5, 6, 9], 'tuple': (83, 12, 3), 'crate': 'dog', 'zeak': 'oh'}
        self.assertEqual(
            '{"Array": [1, 5, 6, 9], "Jack": "jill", "a": 1, "c": 5, "crate": "dog", "pick": "axe", "tuple": [83, 12, 3], "zeak": "oh"}',
            json.dumps(a, item_sort_key=itemgetter(0)))
        self.assertEqual(
            '{"a": 1, "Array": [1, 5, 6, 9], "c": 5, "crate": "dog", "Jack": "jill", "pick": "axe", "tuple": [83, 12, 3], "zeak": "oh"}',
>           json.dumps(a, item_sort_key=lambda kv: kv[0].lower()))
E       AssertionError: '{"a": 1, "Array": [1, 5, 6, 9], "c": 5, "cr[75 chars]oh"}' != '{"Array": [1, 5, 6, 9], "Jack": "jill", "a"[75 chars]oh"}'
E       - {"a": 1, "Array": [1, 5, 6, 9], "c": 5, "crate": "dog", "Jack": "jill", "pick": "axe", "tuple": [83, 12, 3], "zeak": "oh"}
E       ?  --------                                              ----------------
E       + {"Array": [1, 5, 6, 9], "Jack": "jill", "a": 1, "c": 5, "crate": "dog", "pick": "axe", "tuple": [83, 12, 3], "zeak": "oh"}
E       ?                        ++++++++++++++++++++++++

simplejson/tests/test_item_sort_key.py:20: AssertionError
____________________________________________________________ TestItemSortKey.test_simple_first ____________________________________________________________

self = <simplejson.tests.test_item_sort_key.TestItemSortKey testMethod=test_simple_first>

    def test_simple_first(self):
        a = {'a': 1, 'c': 5, 'jack': 'jill', 'pick': 'axe', 'array': [1, 5, 6, 9], 'tuple': (83, 12, 3), 'crate': 'dog', 'zeak': 'oh'}
        self.assertEqual(
            '{"a": 1, "c": 5, "crate": "dog", "jack": "jill", "pick": "axe", "zeak": "oh", "array": [1, 5, 6, 9], "tuple": [83, 12, 3]}',
>           json.dumps(a, item_sort_key=json.simple_first))
E       AssertionError: '{"a": 1, "c": 5, "crate": "dog", "jack": "jill", "p[67 chars] 3]}' != '{"a": 1, "array": [1, 5, 6, 9], "c": 5, "crate": "d[67 chars]oh"}'
E       - {"a": 1, "c": 5, "crate": "dog", "jack": "jill", "pick": "axe", "zeak": "oh", "array": [1, 5, 6, 9], "tuple": [83, 12, 3]}
E       + {"a": 1, "array": [1, 5, 6, 9], "c": 5, "crate": "dog", "jack": "jill", "pick": "axe", "tuple": [83, 12, 3], "zeak": "oh"}

simplejson/tests/test_item_sort_key.py:11: AssertionError
========================================================== 2 failed, 128 passed in 1.01 seconds ===========================================================
```

This is a transcript of my testing:

$ git clone https://github.com/simplejson/simplejson.git
Cloning into 'simplejson'...
remote: Counting objects: 2861, done.
remote: Total 2861 (delta 0), reused 0 (delta 0), pack-reused 2861
Receiving objects: 100% (2861/2861), 1.09 MiB | 362.00 KiB/s, done.
Resolving deltas: 100% (1831/1831), done.
Checking connectivity... done.
$ cd simplejson/
$ python --version
Python 2.7.12
$ virtualenv /tmp/simplejson
New python executable in /tmp/simplejson/bin/python
Installing setuptools, pip, wheel...done.
$ . /tmp/simplejson/bin/activate
(simplejson) $ python --version
Python 2.7.12
(simplejson) $ pip install -e .
Obtaining file:///home/craig/src/github.com/simplejson/simplejson
Installing collected packages: simplejson
  Found existing installation: simplejson 3.11.0
    Not uninstalling simplejson at /home/craig/src/github.com/simplejson/simplejson, outside environment /tmp/simplejson
  Running setup.py develop for simplejson
Successfully installed simplejson
(simplejson) $ nosetests simplejson
S...........................................................................FF.................................................
======================================================================
FAIL: test_case (simplejson.tests.test_item_sort_key.TestItemSortKey)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/craig/src/github.com/simplejson/simplejson/simplejson/tests/test_item_sort_key.py", line 20, in test_case
    json.dumps(a, item_sort_key=lambda kv: kv[0].lower()))
AssertionError: '{"a": 1, "Array": [1, 5, 6, 9], "c": 5, "crate": "dog", "Jack": "jill", "pick": "axe", "tuple": [83, 12, 3], "zeak": "oh"}' != '{"Array": [1, 5, 6, 9], "Jack": "jill", "a": 1, "c": 5, "crate": "dog", "pick": "axe", "tuple": [83, 12, 3], "zeak": "oh"}'

======================================================================
FAIL: test_simple_first (simplejson.tests.test_item_sort_key.TestItemSortKey)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/craig/src/github.com/simplejson/simplejson/simplejson/tests/test_item_sort_key.py", line 11, in test_simple_first
    json.dumps(a, item_sort_key=json.simple_first))
AssertionError: '{"a": 1, "c": 5, "crate": "dog", "jack": "jill", "pick": "axe", "zeak": "oh", "array": [1, 5, 6, 9], "tuple": [83, 12, 3]}' != '{"a": 1, "array": [1, 5, 6, 9], "c": 5, "crate": "dog", "jack": "jill", "pick": "axe", "tuple": [83, 12, 3], "zeak": "oh"}'

----------------------------------------------------------------------
Ran 127 tests in 0.381s

FAILED (SKIP=1, failures=2)
(simplejson) $

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,3 +1,9 @@
+Version 3.11.1 released 2017-06-19
+
+* Fix issue with item_sort_key when speedups are available, and add
+  auto-discovery to test suites to prevent similar regressions
+  https://github.com/simplejson/simplejson/issues/173
+
 Version 3.11.0 released 2017-06-18
 
 * docstring fix in JSONEncoder

--- a/conf.py
+++ b/conf.py
@@ -36,7 +36,7 @@ master_doc = 'index'
 
 # General substitutions.
 project = 'simplejson'
-copyright = '2016, Bob Ippolito'
+copyright = '2017, Bob Ippolito'
 
 # The default replacements for |version| and |release|, also used in various
 # other places throughout the built documents.
@@ -44,7 +44,7 @@ copyright = '2016, Bob Ippolito'
 # The short X.Y version.
 version = '3.11'
 # The full version, including alpha/beta/rc tags.
-release = '3.11.0'
+release = '3.11.1'
 
 # There are two options for replacing |today|: either, you set today to some
 # non-false value, then it is used:

--- a/setup.py
+++ b/setup.py
@@ -11,7 +11,7 @@ from distutils.errors import CCompilerError, DistutilsExecError, \
     DistutilsPlatformError
 
 IS_PYPY = hasattr(sys, 'pypy_translation_info')
-VERSION = '3.11.0'
+VERSION = '3.11.1'
 DESCRIPTION = "Simple, fast, extensible JSON encoder/decoder for Python"
 
 with open('README.rst', 'r') as f:

--- a/simplejson/__init__.py
+++ b/simplejson/__init__.py
@@ -97,7 +97,7 @@ Using simplejson.tool from the shell to validate and pretty-print::
     Expecting property name: line 1 column 3 (char 2)
 """
 from __future__ import absolute_import
-__version__ = '3.11.0'
+__version__ = '3.11.1'
 __all__ = [
     'dump', 'dumps', 'load', 'loads',
     'JSONDecoder', 'JSONDecodeError', 'JSONEncoder',

--- a/simplejson/_speedups.c
+++ b/simplejson/_speedups.c
@@ -2678,17 +2678,17 @@ encoder_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
             if (!item_sort_key)
                 goto bail;
         }
-        if (item_sort_key == Py_None) {
-            Py_INCREF(Py_None);
-            s->item_sort_kw = Py_None;
-        }
-        else {
-            s->item_sort_kw = PyDict_New();
-            if (s->item_sort_kw == NULL)
-                goto bail;
-            if (PyDict_SetItemString(s->item_sort_kw, "key", item_sort_key))
-                goto bail;
-        }
+    }
+    if (item_sort_key == Py_None) {
+        Py_INCREF(Py_None);
+        s->item_sort_kw = Py_None;
+    }
+    else {
+        s->item_sort_kw = PyDict_New();
+        if (s->item_sort_kw == NULL)
+            goto bail;
+        if (PyDict_SetItemString(s->item_sort_kw, "key", item_sort_key))
+            goto bail;
     }
     Py_INCREF(sort_keys);
     s->sort_keys = sort_keys;

--- a/simplejson/tests/__init__.py
+++ b/simplejson/tests/__init__.py
@@ -2,6 +2,7 @@ from __future__ import absolute_import
 import unittest
 import doctest
 import sys
+import os
 
 
 class NoExtensionTestSuite(unittest.TestSuite):
@@ -35,38 +36,13 @@ def additional_tests(suite=None):
 
 def all_tests_suite():
     def get_suite():
+        suite_names = [
+            'simplejson.tests.%s' % (os.path.splitext(f)[0],)
+            for f in os.listdir(os.path.dirname(__file__))
+            if f.startswith('test_') and f.endswith('.py')
+        ]
         return additional_tests(
-            unittest.TestLoader().loadTestsFromNames([
-                'simplejson.tests.test_bitsize_int_as_string',
-                'simplejson.tests.test_bigint_as_string',
-                'simplejson.tests.test_check_circular',
-                'simplejson.tests.test_decode',
-                'simplejson.tests.test_default',
-                'simplejson.tests.test_dump',
-                'simplejson.tests.test_encode_basestring_ascii',
-                'simplejson.tests.test_encode_for_html',
-                'simplejson.tests.test_errors',
-                'simplejson.tests.test_fail',
-                'simplejson.tests.test_float',
-                'simplejson.tests.test_indent',
-                'simplejson.tests.test_iterable',
-                'simplejson.tests.test_pass1',
-                'simplejson.tests.test_pass2',
-                'simplejson.tests.test_pass3',
-                'simplejson.tests.test_recursion',
-                'simplejson.tests.test_scanstring',
-                'simplejson.tests.test_separators',
-                'simplejson.tests.test_speedups',
-                'simplejson.tests.test_str_subclass',
-                'simplejson.tests.test_unicode',
-                'simplejson.tests.test_decimal',
-                'simplejson.tests.test_tuple',
-                'simplejson.tests.test_namedtuple',
-                'simplejson.tests.test_tool',
-                'simplejson.tests.test_for_json',
-                'simplejson.tests.test_subclass',
-                'simplejson.tests.test_raw_json',
-            ]))
+            unittest.TestLoader().loadTestsFromNames(suite_names))
     suite = get_suite()
     import simplejson
     if simplejson._import_c_make_encoder() is None:

--- a/simplejson/tests/test_item_sort_key.py
+++ b/simplejson/tests/test_item_sort_key.py
@@ -18,3 +18,10 @@ class TestItemSortKey(TestCase):
         self.assertEqual(
             '{"a": 1, "Array": [1, 5, 6, 9], "c": 5, "crate": "dog", "Jack": "jill", "pick": "axe", "tuple": [83, 12, 3], "zeak": "oh"}',
             json.dumps(a, item_sort_key=lambda kv: kv[0].lower()))
+
+    def test_item_sort_key_value(self):
+        # https://github.com/simplejson/simplejson/issues/173
+        a = {'a': 1, 'b': 0}
+        self.assertEqual(
+            '{"b": 0, "a": 1}',
+            json.dumps(a, item_sort_key=lambda kv: kv[1]))

